frontend/src/metabase/visualizations/lib/renderer_utils.js

```diff
@@ -223,6 +223,57 @@ function parseTimestampAndWarn(value, unit) {
   return { parsedValue: m };
 }
 
+/*
+
+A waterfall chart is essentially a stacked bar chart.
+It consists of the following (from the topmost):
+- the "green bar" for the positives/increases
+- the "red bar" for the negatives/decreases
+- the "invisible beam" supporting either the green or red bar
+
+Note the green and red bars are mutually exclusive (i.e. if one has
+a positive value, the other is zero). This is done so we need not have
+conditional fill color.
+
+*/
+export function syntheticStackedBarsForWaterfallChart(datas) {
+  const stackedBarsDatas = datas.slice();
+
+  const mainSeries = stackedBarsDatas[0];
+  const mainValues = mainSeries.map(d => d[1]);
+  const totalValue = mainValues.reduce((total, value) => total + value, 0);
+  const total = ["Total", totalValue];
+  // $FlowFixMe cloning for the total bar
+  total._origin = {
+    seriesIndex: mainSeries[0]._origin.seriesIndex,
+    rowIndex: mainSeries.length,
+    cols: mainSeries[0]._origin.cols,
+    row: total,
+  };
+
+  const values = [...mainValues, totalValue];
+  let offset = 0;
+  const beams = [];
+  for (let i = 0; i < values.length - 1; ++i) {
+    beams.push(offset);
+    offset += values[i];
+  }
```

Comment:
Unimportant, but maybe `values.forEach` instead? `i` is only used to index into `values`.

Reply:
Very good point! While at that, I decided to go ahead and use `reduce` which IMO more readable anyway.